### PR TITLE
JS: Model JSON schema validation and 'ajv'

### PR DESCRIPTION
**Pull request by asgerf**
Adds a model of `ajv` along with some general concepts for JSON schema validation.

It has the following impact on our security queries:
- Validating an object against a schema blocks flow of the `TaintedObject` label, preventing FPs from queries like prototype pollution and NoSQL injection.
- Validating a deeply user-controlled object with `allErrors: true` is flagged as potential a denial-of-service attack. For this I introduced the somewhat generically-named query 'Resource exhaustion from deep object traversal'. It's only used for `ajv` at the moment, but I avoided using `ajv` or JSON schema validation as part of its name, so it's easier to extend in the future.
- The error message from `ajv` is treated as a source for ExceptionXss. This query has been rephrased to include errors in a more broad sense, which have not necessarily been through a `throw/catch`.
- Regular expressions in a JSON schema are now analyzed by our RegExp queries, if the JSON schema is written in a JS file. (See internal issue)

Evaluations (internal links):
- [Security on default slugs](https://github.com/dsp-testing/asgerf-bcqs/tree/auto/data/workflow/ajv2-default/reports) looks ok.
  - I've confirmed that the minor regression found in the slowest repo goes away after rebasing on main. In particular, the caching of `getAType` fixed it.
- [Security on ajv slugs](https://github.com/dsp-testing/asgerf-bcqs/tree/auto/data/workflow/ajv3/reports) shows a couple new results.

### CODE REVIEW

**Review by erik-krogh — COMMENTED** [review]
Looks pretty good, I just have some minor comments. 

Do we flag CVE-2020-8192 with the new query?

File: javascript/ql/src/Security/CWE-079/ExceptionXss.ql [hunk 1]
```diff
@@ -15,8 +15,18 @@ import javascript
...
+/**
+ * Gets a description of the source.
+ */
+string getSourceDescription(DataFlow::Node source) {
+  result = source.(ErrorSource).getDescription()
+  or
+  not source instanceof ErrorSource and
+  result = "Exception text"
+}
+
```
[the remark this answers is not in the document]
Comment (asgerf) [reply javascript/ql/src/Security/CWE-079/ExceptionXss.ql]: I've introduced a `Source` class as ExceptionXss which previously just used `Xss::Shared::Source`.

File: javascript/ql/src/semmle/javascript/security/dataflow/DeepObjectResourceExhaustion.qll [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+   */
+  class Configuration extends TaintTracking::Configuration {
+    Configuration() { this = "DeepObjectResourceExhaustion" }
+
+    override predicate isSource(DataFlow::Node source, DataFlow::FlowLabel label) {
+      source instanceof Source and label = TaintedObject::label()
+      or
+      // We currently can't expose the TaintedObject label in the Customizations library
+      // so just add its default sources here.
+      source instanceof TaintedObject::Source and label = TaintedObject::label()
```
[the remark this answers is not in the document]
Comment (asgerf) [reply javascript/ql/src/semmle/javascript/security/dataflow/DeepObjectResourceExhaustion.qll]: Subsumed by other change

File: javascript/ql/src/semmle/javascript/security/dataflow/DeepObjectResourceExhaustion.qll [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      or
+      source instanceof RemoteFlowSource and label.isTaint()
```
[the remark this answers is not in the document]
Comment (asgerf) [reply javascript/ql/src/semmle/javascript/security/dataflow/DeepObjectResourceExhaustion.qll]: As the comment above mentioned, it wasn't possible to refer to `TaintedObject` in a customizations library. But I guess that was just me being lazy, so I went ahead and added a `TaintedObjectCustomizations` file that exposes an abstract flow label for reference within other customizations libraries, but without being materialized.

**Comment by asgerf** [comment]
Thanks @erik-krogh!

Address review comments. I also noticed I forgot to add the tests for the new query, so there's a new commit with tests as well.

> Do we flag CVE-2020-8192 with the new query?

Ah, forgot to mention this, we don't flag this yet.

**Review by erik-krogh — APPROVED** [review]

**Pull request merged**